packages/devtools_app/lib/src/flutter/initializer.dart

```diff
@@ -67,51 +65,51 @@ class _InitializerState extends State<Initializer>
       });
     });
 
+    // If we become disconnected, attempt to reconnect.
     autoDispose(
-      serviceManager.onStateChange.listen((_) {
-        // Generally, empty setState calls in Flutter should be avoided.
-        // However, serviceManager is an implicit part of this state.
-        // This setState call is alerting a change in the serviceManager's
-        // state.
-        setState(() {});
+      serviceManager.onStateChange.where((connected) => !connected).listen((_) {
         // TODO(https://github.com/flutter/devtools/issues/1285): On losing
         // the connection, only provide an option to reconnect; don't
         // immediately go to the connection page.
-        // If we've become disconnected, attempt to reconnect.
-        _navigateToConnectPage();
+        _attemptUrlConnection();
       }),
     );
+    // Trigger a rebuild when the connection becomes available. This is done
+    // by onConnectionAvailable and not onStateChange because we also need
+    // to have queried what type of app this is before we load the UI.
+    autoDispose(
+      serviceManager.onConnectionAvailable.listen((_) => setState(() {})),
+    );
 
-    if (widget.url != null) {
-      _attemptUrlConnection();
-    } else {
-      _navigateToConnectPage();
-    }
+    _attemptUrlConnection();
   }
 
   Future<void> _attemptUrlConnection() async {
+    if (widget.url == null) {
+      _navigateBackToConnectPage();
+      return;
+    }
+
     final uri = normalizeVmServiceUri(widget.url);
     final connected = await FrameworkCore.initVmService(
       '',
       explicitUri: uri,
       errorReporter: (message, error) =>
           Notifications.of(context).push('$message, $error'),
     );
+
     if (!connected) {
```

Comment:
not your bug but for some defensive programming, maybe wrape the initVmService call with a 
try {
} finally {
  if (!connected) ..
}

Reply:
There's already error handling inside `initVmService` which reports errors as notifications and then returns false. I can add it here, but I think it'd only be hit if the other error handling was removed (which would then effect the connect screen and others). Is it still worth adding here?